sextant/src/main.rs

```diff
@@ -66,9 +75,360 @@ enum Opt {
+                &tmp_root_dir.join(arc_spec_file.file_name().unwrap()),
+            )?;
+            // fs_extra::dir::copy(&arc_spec_file, &tmp_root_dir, &options)?;
+        }
+    }
+
+    println!("find {}:", tmp_dir.path().as_os_str().to_str().unwrap());
+    println!(
+        "{}",
+        String::from_utf8_lossy(
+            &Command::new("find")
+                .arg(tmp_dir.path().as_os_str().to_str().unwrap())
+                .output()?
+                .stdout
+        )
+    );
+
+    // write manifest
+    log::debug!("write manifest");
+    let tmp_manifest_dir = tmp_dir.path().join("manifest").with_extension("yaml");
+    log::debug!(
+        "create file {}",
+        tmp_manifest_dir.as_os_str().to_str().unwrap()
+    );
+    let tmp_manifest_file = File::create(&tmp_manifest_dir)?;
+    log::debug!(
+        "writing file {}",
+        tmp_manifest_dir.as_os_str().to_str().unwrap()
+    );
+    serde_yaml::to_writer(tmp_manifest_file, &manifest)?;
+
+    // remove existing containers
+    // TODO: remove all {registry}/#{name}-#{arch}-* directories
+
+    let npk_dir = registry_dir
+        .join(format!("{}-{}-{}", manifest.name, arch, manifest.version))
+        .with_extension("npk");
+    let fsimg_path = &tmp_dir.path().join("fs").with_extension("img");
+
+    /* The list of pseudo files is target specific.
+     * Add /lib and lib64 on Linux systems.
+     * Add /system on Android. */
+    let mut pseudo_files = vec![
+        ("/tmp", 444),
+        ("/proc", 444),
+        ("/dev", 444),
+        ("/sys", 444),
+        ("/data", 777),
+    ];
+    if arch == "aarch64-unknown-linux-gnu" || arch == "x86_64-unknown-linux-gnu" {
+        pseudo_files.push(("/lib", 444));
+        pseudo_files.push(("/lib64", 444));
+    } else if arch == "aarch64-linux-android" {
+        pseudo_files.push(("/system", 444));
+    }
+
+    // create filesystem image
+    let squashfs_comp = match std::env::consts::OS {
+        "linux" => "gzip",
+        _ => "zstd",
+    };
+    if fs_type == &FsType::SQUASHFS {
+        let mut cmd = Command::new("mksquashfs");
+        cmd.arg(tmp_root_dir.as_os_str().to_str().unwrap())
+            .arg(fsimg_path.as_os_str().to_str().unwrap())
+            .arg("-all-root")
+            .arg("-comp")
+            .arg(squashfs_comp)
+            .arg("-no-progress")
+            .arg("-info");
+        for pseudo_file in pseudo_files {
+            cmd.arg("-p");
+            cmd.arg(format!(
+                "{} d {} {} {}",
+                pseudo_file.0,
+                pseudo_file.1.to_string(),
+                uid,
+                gid
+            ));
+        }
+        dbg!(&cmd);
+        let cmd_output = cmd.output()?;
+        println!("status={}", cmd_output.status);
+        println!("stdout={}", String::from_utf8_lossy(&cmd_output.stdout));
+        println!("stderr={}", String::from_utf8_lossy(&cmd_output.stderr));
+    } else if fs_type == &FsType::EXT4 {
```

Comment:
Skip `ext4` for now. Movement clearly goes into the direction of squashfs. The ext4 support is barely tested and probably removed and added in a clean way when we need it (maybe);